llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -3165,6 +3165,26 @@ Instruction *InstCombinerImpl::foldICmpAddConstant(ICmpInst &Cmp,
                         Builder.CreateAdd(X, ConstantInt::get(Ty, *C2 - C - 1)),
                         ConstantInt::get(Ty, ~C));
 
+  // zext(V) + C2 <u C -> V + trunc(C2) <u trunc(C) iff C2 s<0 && C s>0
```

Comment:
Think the 'iff ...' is pretty misleading given that you also need 'C2' and 'C' to fit in the new bitwidth.